tests/bsim/bluetooth/host/att/sequential/tester/src/main.c

```diff
@@ -0,0 +1,715 @@
+
+	set_param.min_interval = sys_cpu_to_le16(interval);
+	set_param.max_interval = sys_cpu_to_le16(interval);
+	set_param.channel_map = 0x07;
+	set_param.filter_policy = BT_LE_ADV_FP_NO_FILTER;
+	set_param.type = BT_HCI_ADV_IND;
+	set_param.own_addr_type = 0x01; /* random */
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_ADV_PARAM, sizeof(set_param));
+	__ASSERT_NO_MSG(buf);
+	net_buf_add_mem(buf, &set_param, sizeof(set_param));
+
+	send_cmd(BT_HCI_OP_LE_SET_ADV_PARAM, buf, NULL);
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_ADV_ENABLE, 1);
+	__ASSERT_NO_MSG(buf);
+
+	net_buf_add_u8(buf, BT_HCI_LE_ADV_ENABLE);
+	send_cmd(BT_HCI_OP_LE_SET_ADV_ENABLE, buf, NULL);
+}
+
+NET_BUF_POOL_DEFINE(acl_tx_pool, 5, BT_L2CAP_BUF_SIZE(200), 8, NULL);
+
+struct net_buf *alloc_l2cap_pdu(void)
+{
+	struct net_buf *buf;
+	uint16_t reserve;
+
+	buf = net_buf_alloc(&acl_tx_pool, K_FOREVER);
+	__ASSERT_NO_MSG(buf);
+
+	reserve = sizeof(struct bt_l2cap_hdr);
+	reserve += sizeof(struct bt_hci_acl_hdr) + BT_BUF_RESERVE;
+
+	net_buf_reserve(buf, reserve);
+
+	return buf;
+}
+
+static int send_acl(struct net_buf *buf)
+{
+	struct bt_hci_acl_hdr *hdr;
+	uint8_t flags = BT_ACL_START_NO_FLUSH;
+
+	hdr = net_buf_push(buf, sizeof(*hdr));
+	hdr->handle = sys_cpu_to_le16(bt_acl_handle_pack(conn_handle, flags));
+	hdr->len = sys_cpu_to_le16(buf->len - sizeof(*hdr));
+
+	bt_buf_set_type(buf, BT_BUF_ACL_OUT);
+
+	k_sem_take(&acl_pkts, K_FOREVER);
+
+	return bt_send(buf);
+}
+
+static void send_l2cap_packet(struct net_buf *buf, uint16_t cid)
+{
+	struct bt_l2cap_hdr *hdr;
+
+	hdr = net_buf_push(buf, sizeof(*hdr));
+	hdr->len = sys_cpu_to_le16(buf->len - sizeof(*hdr));
+	hdr->cid = sys_cpu_to_le16(cid);
+
+	/* Always entire packets, no HCI fragmentation */
+	__ASSERT_NO_MSG(buf->len <= CONFIG_BT_BUF_ACL_TX_SIZE);
+
+	send_acl(buf);
+}
+
+static void gatt_write(bool req)
+{
+	static uint8_t data[] = "write";
+	uint16_t handle = server_write_handle;
+	struct net_buf *buf = alloc_l2cap_pdu();
+
+	net_buf_add_u8(buf, req ?
+		       BT_ATT_OP_WRITE_REQ : BT_ATT_OP_WRITE_CMD);
+
+	net_buf_add_le16(buf, handle);
+	net_buf_add_mem(buf, data, sizeof(data));
+
+	LOG_INF("send ATT write %s", req ? "REQ" : "CMD");
+	send_l2cap_packet(buf, BT_L2CAP_CID_ATT);
+}
+
+static void gatt_notify(void)
+{
+	static uint8_t data[] = "notification";
+	uint16_t handle = HVX_HANDLE;
+	struct net_buf *buf = alloc_l2cap_pdu();
+
+	net_buf_add_u8(buf, BT_ATT_OP_NOTIFY);
+	net_buf_add_le16(buf, handle);
+	net_buf_add_mem(buf, data, sizeof(data));
+
+	LOG_INF("send ATT notification");
+	send_l2cap_packet(buf, BT_L2CAP_CID_ATT);
+}
+
+static void gatt_indicate(void)
+{
+	static uint8_t data[] = "indication";
+	uint16_t handle = HVX_HANDLE;
+	struct net_buf *buf = alloc_l2cap_pdu();
+
+	net_buf_add_u8(buf, BT_ATT_OP_INDICATE);
+	net_buf_add_le16(buf, handle);
+	net_buf_add_mem(buf, data, sizeof(data));
+
+	LOG_INF("send ATT indication");
+	send_l2cap_packet(buf, BT_L2CAP_CID_ATT);
+}
+
+static void init_ll(void)
```

Comment:
This function name is a bit misleading, as it does not initialize the LL, but rather sends a few commands to the controller.